lsp/nls/src/linearization/mod.rs

```diff
@@ -210,12 +210,27 @@ impl<'a> Linearizer for AnalysisHost<'a> {
                         index: id_gen.get_and_advance(),
                     };
                     self.env.insert(ident.to_owned(), id);
+                    let (pos, kind) = match term {
+                        Term::LetPattern(..) => (
+                            ident.pos,
+                            TermKind::Declaration(ident.to_owned(), Vec::new(), value_ptr),
+                        ),
+                        Term::FunPattern(..) => (
+                            pos,
+                            TermKind::Declaration(
+                                ident.to_owned(),
+                                Vec::new(),
+                                ValueState::Unknown,
+                            ),
+                        ),
+                        _ => unreachable!(),
+                    };
                     lin.push(LinearizationItem {
                         env: self.env.clone(),
                         id,
                         ty,
-                        pos: ident.pos,
-                        kind: TermKind::Declaration(ident.to_owned(), Vec::new(), value_ptr),
```

Comment:
Are you sure this code was handling the function item itself? It seems it added a `TermKind::Declaration` item in the linearization. Something used e.g. for the goto definition feature, which I believe is different from the linearized term itself. In particular, in the case of the let-binding, you still use the position of the indent, which is weird.

Wouldn't the function be linearized when the body is walked/typechecked?

Reply:
Yes it was. It represents the function argument as a `TermKind::Declaration(..)`, so it can be used in the body. 

> In particular, in the case of the let-binding, you still use the position of the indent, which is weird.

Maybe, but I guess this is done because when you hover on the `ident` you get the type and other information, and  we don't want this information if we hover on *any* other part of the let binding. I think this was the rationale for using the ident's position for the function's position. For the function's case, I'm thinking we can create a `TermKind` for the `ident` , and another for the function itself.

Reply:
>So the previous code was kinda correct, isn't it? What are the cases that are problematic? When you try to hover before the body, like on the `=>` ?

Yes, hovering on the `=>` would be the problematic case. Or maybe the `fun` keyword.

>But then, this is the same thing with the function? You don't want to get the declaration info when hovering on the body, do you ? I might be misunderstanding something, but from what I understand, shouldn't we rather keep the declaration item, and add before that an item for the whole function, if that's what's missing?

Yes, we actually keep the declaration item for both cases, the only thing we change is the value the declaration item points to for the function case. Previously, for the function case, we create a declaration item that points back to itself; I changed it to have an unknown value.

Reply:
I see, thanks for the clarification. I'm still not sure to understand what's the answer to this question though:

> shouldn't we rather keep the declaration item, and add before that an item for the whole function, if that's what's missing?

Reply:
Yes, exactly. I was even considering creating a separate `TermKind` for functions.

Reply:
So should we do that instead? And keep the term declaration position to be the one of the variable? Because it seems a bit strange to affect the position of the whole function for the declaration. The declaration usually correspond to the introduction of an identifier

Reply:
For example, currently, when you hover on the parameter of a function within a typed block, you get the type of this argument, which is arguably useful. Is this behavior preserved by this PR?

Reply:
> So should we do that instead? And keep the term declaration position to be the one of the variable? Because it seems a bit strange to affect the position of the whole function for the declaration. The declaration usually correspond to the introduction of an identifier

Yes, we should do that. But instead of a new term kind for functions, we can reuse `TermKind::Structure` , which represents an arbitrary term.

Reply:
> For example, currently, when you hover on the parameter of a function within a typed block, you get the type of this argument, which is arguably useful. Is this behavior preserved by this PR?

Currently, it doesn't preserve this. But with the suggestion above, it preserves it.

Reply:
Yes, this sounds good. Do you want to fix that here or in a new PR?

Reply:
I'll fix it here